File: core/effects/edge_detect.py
"""
bitboylab_boy.core.effects.edge_detect
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Sobel-based edge detection rendered as single-color line art.

Algorithm
---------
1. Convert image to grayscale.
2. Compute X and Y gradients via vectorized NumPy slicing (Sobel-like).
3. Compute gradient magnitude: √(Gx² + Gy²), normalized to [0, 255].
4. Threshold the magnitude map to boolean edge mask.
5. Optionally dilate the edge mask for thicker lines.
6. Return an RGBA image — edge pixels fully opaque, others transparent.

Performance
-----------
Steps 2–4 are fully vectorized using NumPy advanced slicing — no Python
loops over pixels. Dilation in step 5 uses a simple boolean OR shift
loop (O(thickness × W × H)) — acceptable for thickness ≤ 10.
"""
from __future__ import annotations

import numpy as np
from PIL import Image

from utils.color import hex_to_rgb


def apply_edge_detect(
    img: Image.Image,
    mask: np.ndarray,
    sensitivity: int = 50,
    thickness: int = 1,
    color_hex: str = "#ffffff",
) -> Image.Image:
    """Detect edges via Sobel operators and render them as line art.

    Args:
        img:         Source image (any mode).
        mask:        Boolean NumPy array (H, W) — edges only drawn here.
        sensitivity: 0–100. Higher = more edges detected.
                     Internally mapped as threshold = (101 - sensitivity) × 2.55.
        thickness:   Line thickness in pixels (1–10).
                     Implemented via iterative boolean dilation.
        color_hex:   Hex color string for the edge lines.

    Returns:
        RGBA image — edges in *color_hex* fully opaque, background transparent.
    """
    rgb_color = np.array(hex_to_rgb(color_hex), dtype=np.uint8)
    arr = np.array(img.convert("L")).astype(np.float32)
    h, w = arr.shape

    # --- Vectorized Sobel gradient (no per-pixel loop) ---
    gx = np.zeros_like(arr)
    gy = np.zeros_like(arr)
    gx[:, 1:-1] = arr[:, 2:] - arr[:, :-2]    # horizontal gradient
    gy[1:-1, :] = arr[2:, :] - arr[:-2, :]    # vertical gradient

    mag = np.hypot(gx, gy)  # equivalent to sqrt(gx² + gy²)
    if mag.max() > 0:
        mag = mag / mag.max() * 255.0

    # Map sensitivity [0, 100] → threshold [255, 0]
    threshold = (101 - sensitivity) * 2.55
    edges: np.ndarray = (mag > threshold) & mask

    # --- Dilation for thicker lines (iterative boolean OR shift) ---
    if thickness > 1:
        for _ in range(thickness - 1):
            dilated = edges.copy()
            dilated[1:] |= edges[:-1]    # shift down
            dilated[:-1] |= edges[1:]   # shift up
            dilated[:, 1:] |= edges[:, :-1]  # shift right
            dilated[:, :-1] |= edges[:, 1:]  # shift left
            edges = dilated

    # --- Build RGBA output ---
    out = np.zeros((h, w, 4), dtype=np.uint8)
    out[edges] = [*rgb_color, 255]  # edge pixels: color + fully opaque
    return Image.fromarray(out, "RGBA")

File: frontend/src/lib/effects/edge_detect.ts
import { LayerConfig } from "../../types";
import { hexToRgb } from "../color";
import { getLuminance } from "../mask";

/**
 * Replicates bitboylab_boy.core.effects.edge_detect.
 * Sobel-based edge detection.
 */
export function applyEdgeDetect(
    sourceImageData: ImageData,
    layer: LayerConfig
): ImageData {
    const sensitivity = layer.sensitivity !== undefined ? Number(layer.sensitivity) : 50;
    const thickness = layer.thickness !== undefined ? Number(layer.thickness) : 1;
    const colorHex = layer.color as string || "#ffffff";
    const [rStr, gStr, bStr] = hexToRgb(colorHex);

    const { width, height, data } = sourceImageData;
    const outData = new Uint8ClampedArray(width * height * 4);

    // 1. Convert to Grayscale (1D array for easier index math)
    const gray = new Float32Array(width * height);
    for (let i = 0, j = 0; i < data.length; i += 4, j++) {
        gray[j] = getLuminance(data[i], data[i + 1], data[i + 2]);
    }

    // 2. Sobel Gradient & Magnitude
    // We compute gradients skipping the 1px border like the Python numpy advanced slicing
    const mag = new Float32Array(width * height);
    let maxMag = 0;

    for (let y = 1; y < height - 1; y++) {
        for (let x = 1; x < width - 1; x++) {
            const idx = y * width + x;
            // Simplified Sobel / Cross differences
            const gx = gray[idx + 1] - gray[idx - 1];
            const gy = gray[idx + width] - gray[idx - width];
            
            const currentMag = Math.sqrt(gx * gx + gy * gy);
            mag[idx] = currentMag;
            if (currentMag > maxMag) maxMag = currentMag;
        }
    }

    // Normalize magnitude to 0-255 if maxMag > 0
    if (maxMag > 0) {
        for (let i = 0; i < mag.length; i++) {
            mag[i] = (mag[i] / maxMag) * 255.0;
        }
    }

    // 3. Thresholding
    // Map sensitivity [0, 100] -> threshold [255, 0]
    const threshold = (101 - sensitivity) * 2.55;
    let edges = new Uint8Array(width * height);
    
    for (let i = 0; i < mag.length; i++) {
        edges[i] = mag[i] > threshold ? 1 : 0;
    }

    // 4. Dilation for thickness
    if (thickness > 1) {
        for (let t = 0; t < thickness - 1; t++) {
            const dilated = new Uint8Array(edges);
            for (let y = 0; y < height; y++) {
                for (let x = 0; x < width; x++) {
                    const idx = y * width + x;
                    if (edges[idx] === 1) {
                        if (x < width - 1) dilated[idx + 1] = 1; // shift right
                        if (x > 0) dilated[idx - 1] = 1; // shift left
                        if (y < height - 1) dilated[idx + width] = 1; // shift down
                        if (y > 0) dilated[idx - width] = 1; // shift up
                    }
                }
            }
            edges = dilated;
        }
    }

    // 5. Build RGBA Output
    for (let i = 0, j = 0; i < data.length; i += 4, j++) {
        if (edges[j] === 1) {
            outData[i] = rStr;
            outData[i + 1] = gStr;
            outData[i + 2] = bStr;
            outData[i + 3] = 255; // Fully opaque edges
        } else {
            outData[i] = 0;
            outData[i + 1] = 0;
            outData[i + 2] = 0;
            outData[i + 3] = 0; // Transparent background
        }
    }

    return new ImageData(outData, width, height);
}
